### PR TITLE
ANSI `AT TIME ZONE` parsing improvements

### LINKED ISSUE #3111: Issues in Redshift Dialect
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Redshift specific functions are causing the isssues in applying the rules. When I commented the getdate function lines, I could see the format working .
Settings:

```
[sqlfluff]
dialect = redshift
[sqlfluff:rules]
tab_space_size = 4
comma_style = leading
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all
capitalisation_policy = upper
extended_capitalisation_policy = upper
[sqlfluff:rules:L003]
indent_unit = space
tab_space_size = 4
[sqlfluff:rules:L004]
indent_unit = space
[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = upper
[sqlfluff:rules:L014]
# Keywords
extended_capitalisation_policy = upper
[sqlfluff:rules:L019]
# Keywords
comma_style = leading
```

sql:
```sql
insert INTO test1
(
	col1
    ,col2
    ,col3
    ,col4
    ,col5
    ,col6
)
select
    STG.col1
    ,COALESCE(LEAD(STG.EFFECTIVE_START_TIMESTAMP - 1) OVER (PARTITION BY STG.col1, STG.col2 ORDER BY STG.EFFECTIVE_START_TIMESTAMP),'9999-12-31' ) AS EFFECTIVE_END_TIMESTAMP 	 						
	,LEAD(STG.col3) OVER(PARTITION BY STG.col1, STG.col2 ORDER BY STG.EFFECTIVE_START_TIMESTAMP) AS COL2
	,(GETDATE() AT TIME ZONE 'UTC') AT TIME ZONE 'AEST'   AS ROW_CREATED_TIMESTAMP
    ,((GETDATE() AT TIME ZONE 'UTC') AT TIME ZONE 'AEST')   AS ROW_UPDATED_TIMESTAMP
    ,CASE WHEN EFFECTIVE_END_TIMESTAMP = '9999-12-31' THEN 1 ELSE 0 END AS CURRENT_IND   
FROM table1 STG
;
```

### Expected Behaviour

I am expecting the output to be like below:

```sql
INSERT INTO TEST1
(
    COL1
    , COL2
    , COL3
    , COL4
    , COL5
    , COL6
)
SELECT
    TABLE1.COL1
    , COALESCE(
        LEAD(
            TABLE1.EFFECTIVE_START_TIMESTAMP - 1
        ) OVER (
            PARTITION BY
                TABLE1.COL1, TABLE1.COL2
            ORDER BY TABLE1.EFFECTIVE_START_TIMESTAMP
        )
        , '9999-12-31'
    ) AS EFFECTIVE_END_TIMESTAMP
    , LEAD(
        TABLE1.COL3
    ) OVER(
        PARTITION BY
            TABLE1.COL1, TABLE1.COL2
        ORDER BY TABLE1.EFFECTIVE_START_TIMESTAMP
    ) AS COL2
    ,(GETDATE() AT TIME ZONE 'UTC') AT TIME ZONE 'AEST'   AS ROW_CREATED_TIMESTAMP
    ,((GETDATE() AT TIME ZONE 'UTC') AT TIME ZONE 'AEST')   AS ROW_UPDATED_TIMESTAMP
    , CASE
        WHEN TABLE1.EFFECTIVE_END_TIMESTAMP = '9999-12-31' THEN 1 ELSE 0
    END AS CURRENT_IND
FROM TABLE1;
```

### Observed Behaviour

```sql
insert INTO test1
(
	col1
    ,col2
    ,col3
    ,col4
    ,col5
    ,col6
)
select
    STG.col1
    ,COALESCE(LEAD(STG.EFFECTIVE_START_TIMESTAMP - 1) OVER (PARTITION BY STG.col1, STG.col2 ORDER BY STG.EFFECTIVE_START_TIMESTAMP),'9999-12-31' ) AS EFFECTIVE_END_TIMESTAMP 	 						
	,LEAD(STG.col3) OVER(PARTITION BY STG.col1, STG.col2 ORDER BY STG.EFFECTIVE_START_TIMESTAMP) AS COL2
	,(GETDATE() AT TIME ZONE 'UTC') AT TIME ZONE 'AEST'   AS ROW_CREATED_TIMESTAMP
    ,((GETDATE() AT TIME ZONE 'UTC') AT TIME ZONE 'AEST')   AS ROW_UPDATED_TIMESTAMP
    ,CASE WHEN EFFECTIVE_END_TIMESTAMP = '9999-12-31' THEN 1 ELSE 0 END AS CURRENT_IND   
FROM table1 STG
;
```

### How to reproduce

I am on latest version of sqlfluff and can easily replicated using the above mentioned code.

### Dialect

redshift

### Version

sqlfluff, version 0.12.0

### Configuration

setup.cfg

```
[sqlfluff]
dialect = redshift
[sqlfluff:rules]
tab_space_size = 4
comma_style = leading
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all
capitalisation_policy = upper
extended_capitalisation_policy = upper
[sqlfluff:rules:L003]
indent_unit = space
tab_space_size = 4
[sqlfluff:rules:L004]
indent_unit = space
[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = upper
[sqlfluff:rules:L014]
# Keywords
extended_capitalisation_policy = upper
[sqlfluff:rules:L019]
# Keywords
comma_style = leading
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made

Fixes #3111 

### Are there any other side effects of this change that we should be aware of?
Shouldn't be, but it is a change to our core expression grammar.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_bigquery.py [hunk 1]
```diff
@@ -488,7 +488,6 @@ class FunctionSegment(ansi.FunctionSegment):
                     Ref("DatetimeUnitSegment"),
                     "FROM",
                     Ref("ExpressionSegment"),
-                    Ref("TimeZoneGrammar", optional=True),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_bigquery.py:491]: This is no longer required now we're adding it to the core expression grammar.

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 2]
```diff
@@ -362,7 +362,6 @@
     ColumnReferenceSegment=Sequence(
         ansi.ColumnReferenceSegment,
         Ref("ArrayAccessorSegment", optional=True),
-        Ref("TimeZoneGrammar", optional=True),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_postgres.py:365]: This is no longer required now we're adding it to the core expression grammar.

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 3]
```diff
@@ -515,13 +514,12 @@ class DateTimeTypeIdentifier(BaseSegment):
...
 
 class DateTimeLiteralGrammar(BaseSegment):
-    """Literal Date Time with optional casting to Time Zone."""
+    """Literal Date Time."""
 
     type = "datetime_literal"
     match_grammar = Sequence(
         Ref("DateTimeTypeIdentifier"),
         Ref("QuotedLiteralSegment"),
-        Ref("TimeZoneGrammar", optional=True),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_postgres.py:524]: This is no longer required now we're adding it to the core expression grammar.

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 4]
```diff
@@ -541,10 +539,7 @@ class DatatypeSegment(ansi.DatatypeSegment):
         ),
         OneOf(
             Ref("WellKnownTextGeometrySegment"),
-            Sequence(
-                Ref("DateTimeTypeIdentifier"),
-                Ref("TimeZoneGrammar", optional=True),
-            ),
+            Ref("DateTimeTypeIdentifier"),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_postgres.py:542]: This is no longer required now we're adding it to the core expression grammar.

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 5]
```diff
@@ -163,7 +163,6 @@
         Sequence(
             ansi.ColumnReferenceSegment,
             AnyNumberOf(Ref("ArrayAccessorSegment")),
-            Ref("TimeZoneGrammar", optional=True),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_redshift.py:166]: This is no longer required now we're adding it to the core expression grammar.

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 6]
```diff
@@ -358,10 +357,7 @@ class DatatypeSegment(BaseSegment):
             "BPCHAR",
             "TEXT",
         ),
-        Sequence(
-            Ref("DateTimeTypeIdentifier"),
-            Ref("TimeZoneGrammar", optional=True),
-        ),
+        Ref("DateTimeTypeIdentifier"),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_redshift.py:360]: This is no longer required now we're adding it to the core expression grammar.

**Review by barrywhart — COMMENTED** [review]
Should we add a test case similar to the SQL provided by the user, the one where the resulting formatting was wrong?

**Comment by tunetheweb** [comment]
I don’t think we need to. The basic issue boiled down to “why aren’t the capitalisation rules being applied to my SQL” and the answer was “we don’t apply ANY rules on code with parsing errors”.

**Comment by barrywhart** [comment]
Oh, it wasn't parsing at all? I hadn't noticed that.

**Review by barrywhart — APPROVED** [review]

**Pull request merged**